pkg/acquisition/modules/kinesis/kinesis.go

```diff
@@ -0,0 +1,473 @@
+		k.logger.Infof("Waiting for consumer registration %d", i)
+	}
+	return fmt.Errorf("consumer %s is not active after %d tries", consumerARN, maxTries)
+}
+
+func (k *KinesisSource) RegisterConsumer() (*kinesis.RegisterStreamConsumerOutput, error) {
+	k.logger.Infof("Registering consumer %s", k.Config.ConsumerName)
+	streamConsumer, err := k.kClient.RegisterStreamConsumer(&kinesis.RegisterStreamConsumerInput{
+		ConsumerName: aws.String(k.Config.ConsumerName),
+		StreamARN:    aws.String(k.Config.StreamARN),
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot register stream consumer")
+	}
+	err = k.WaitForConsumerRegistration(*streamConsumer.Consumer.ConsumerARN)
+	if err != nil {
+		return nil, errors.Wrap(err, "Timeout while waiting for consumer to be active")
+	}
+	return streamConsumer, nil
+}
+
+func (k *KinesisSource) ParseAndPushRecords(records []*kinesis.Record, out chan types.Event, logger *log.Entry) {
+	for _, record := range records {
+		var data []CloudwatchSubscriptionLogEvent
+		var err error
+		if k.Config.FromSubscription {
+			//The AWS docs says that the data is base64 encoded
+			//but apparently GetRecords decodes it for us ?
+			data, err = k.decodeFromSubscription(record.Data)
+			if err != nil {
+				logger.Errorf("Cannot decode data: %s", err)
+			}
+		} else {
+			data = []CloudwatchSubscriptionLogEvent{{Message: string(record.Data)}}
+		}
+		for _, event := range data {
+			logger.Infof("got record %s", event.Message)
+			l := types.Line{}
+			l.Raw = event.Message
+			l.Labels = k.Config.Labels
+			l.Time = time.Now()
+			l.Process = true
+			l.Module = k.GetName()
+			//linesRead.With(prometheus.Labels{"source": j.src}).Inc()
+			evt := types.Event{Line: l, Process: true, Type: types.LOG, ExpectMode: leakybucket.LIVE}
+			out <- evt
+		}
+	}
+}
+
+func (k *KinesisSource) ReadFromSubscription(reader kinesis.SubscribeToShardEventStreamReader, out chan types.Event, shardId string, streamName string) error {
+	logger := k.logger.WithFields(log.Fields{"shard_id": shardId})
+	//ghetto sync, kinesis allows to subscribe to a closed shard, which will make the goroutine exit immediately
+	//and we won't be able to start a new one if this is the first one started by the tomb
+	//TODO: look into parent shards to see if a shard is closed before starting to read it ?
+	time.Sleep(time.Second)
+	for {
+		select {
+		case <-k.shardReaderTomb.Dying():
+			logger.Infof("Subscribed shard reader is dying")
+			err := reader.Close()
+			if err != nil {
+				return errors.Wrap(err, "Cannot close kinesis subscribed shard reader")
+			}
+			return nil
+		case event, ok := <-reader.Events():
+			if !ok {
+				logger.Infof("Event chan has been closed")
+				return nil
+			}
+			switch event := event.(type) {
+			case *kinesis.SubscribeToShardEvent:
+				k.ParseAndPushRecords(event.Records, out, logger)
+			case *kinesis.SubscribeToShardEventStreamUnknownEvent:
+				logger.Infof("got an unknown event, what to do ?")
```

Comment:
good question :D

Reply:
I think we will leave it like this for now, I've been unable to find any documentation in which conditions this might happen.